c10/core/TensorImpl.h

```diff
@@ -1438,7 +1438,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   bool is_non_overlapping_and_dense() const {
-    return is_non_overlapping_and_dense_;
+    return support_as_strided() && is_non_overlapping_and_dense_;
```

Comment:
I don't understand this change.

If you're trying to enforce an invariant that `is_non_overlapping_and_dense_` is only true if `support_as_strided()`, then the test `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(!(!support_as_strided() && is_non_overlapping_and_dense_))` seems better.

Reply:
Originally was trying to hide this check from call sites. But yeah it makes more sense to be an check.

Reply:
No, I'm even more confused now.

In my previous comment, what I was trying to establish is an internal invariant on Tensor irrespective of the use sites, that whenever !support_as_strided(), then is_non_overlapping_dense_ is false. When this invariant holds, you DON'T need to put a test in the call site: when is_non_overlapping_and_dense is true, then you can also infer that support_as_strided() is true (one verifies the other). Can we try to establish that invariant? That will lead to a cleaner internal representation and also help @ngimel with devirtualization.

Reply:
Yeah, in #49303 I propose that all TensorImpl derivative classes correctly set bitfields for `is_contiguous` and `is_non_overlapping_and_dense` so that those calls don't need to be virtual. `support_as_strided()` currently is super misleading - it returns that tensor is not XLA, but OpaqueTensorImpl and SparseTensorImpl and possibly BatchedTensorImpl also don't support as strided. 
So, to recap, I agree with @ezyang that we should try to establish invariant that all Impls that don't know how to deal with the concept of `is_contiguous` or `is_non_overlapping_and_dense` should set those bitfields to false, and then the function calls won't have to be guarded on anything.

Reply:
@ngimel @ezyang hmmm I'm fine with setting is_non_overlapping_and_dense to false on XLATensorImpl. (need to double check on is_contiguous). 
Another quick thought: this invariant seem to impose an invariant like
"if you don't support as_strided, set these bitfields to false.". And to me it seems this list of bitfields to set is growing over time, but all of these fields are actually try to represent the memory_format information. 
Alternatively is it better that we have a bitfield for support_as_strided so that every TensorImpl can set, and then is_contiguous/is_non_overlapping_and_dense can just be
```
  return support_as_strided && <check memory format>
```
What do you think?

Reply:
Setting `support_as_strided` as a bitfield makes sense to me, but we don't have available bitfields left, so that would mean increasing TensorImpl size?

Reply:
We have more bitfields left. Due to alignment concerns TensorImpl will always be rounded to a machine word, which means you get 64 bit fields per word.

That being said, I think we should be deliberate and holistic about what bit fields we add to TensorImpl. Every new computed bit field is another place for invariants to fail you. We need to clearly describe what these invariants are, and make it difficult for people to accidentally break them. This is not a small PR, it is going through all of these fields and methods with a fine toothed comb.